packages/SwingSet/test/device-plugin/test-device.js

```diff
@@ -0,0 +1,109 @@
+/* global harden */
+import '@agoric/install-ses';
+import test from 'ava';
+import { buildVatController } from '../../src/index';
+import { buildBridge } from '../../src/devices/bridge';
+import { buildPlugin } from '../../src/devices/plugin';
+
+test.before('initialize hostStore', t => {
+  const m = new Map();
+  const storage = harden({
+    delete(key) {
+      return m.delete(key);
+    },
+    get(key) {
+      return m.get(key);
+    },
+    has(key) {
+      return m.has(key);
+    },
+    set(key, value) {
+      return m.set(key, value);
+    },
+    getKeys() {
+      return [...m.keys()];
+    },
+  });
+  t.context.hostStorage = storage;
+});
+
+const setupVatController = async t => {
+  const inputQueue = [];
+  const queueThunkForKernel = async thunk => {
+    inputQueue.push(thunk);
+  };
+
+  const pluginRequire = mod => {
+    t.is(mod, 'pingpong');
+    // eslint-disable-next-line global-require
+    return require('./pingpong');
+  };
+  const plugin = buildPlugin(__dirname, pluginRequire, queueThunkForKernel);
+  const bridge = buildBridge();
+  const config = {
+    bootstrap: 'bootstrap',
+    vats: {
+      bootstrap: {
+        sourceSpec: require.resolve('./bootstrap'),
+      },
+      bridge: {
+        sourceSpec: require.resolve('./vat-bridge'),
+      },
+    },
+    devices: [
+      ['plugin', plugin.srcPath, plugin.endowments],
+      ['bridge', bridge.srcPath, bridge.endowments],
+    ],
+  };
+  const c = await buildVatController(config, ['plugin'], {
+    hostStorage: t.context.hostStorage,
+  });
+  const cycle = async () => {
+    await c.run();
+    while (inputQueue.length) {
+      inputQueue.shift()();
+      // eslint-disable-next-line no-await-in-loop
+      await c.run();
+    }
+  };
+  return { bridge, cycle, dump: c.dump, plugin, queueThunkForKernel };
+};
+
+test.serial('plugin first time', async t => {
+  const { bridge, cycle, dump, queueThunkForKernel } = await setupVatController(
+    t,
+  );
+
+  queueThunkForKernel(() => bridge.deliverInbound('pingpong'));
+  await cycle();
+
+  t.deepEqual(dump().log, [
+    'starting plugin test',
+    'installing pingPongP',
+    'starting pingpong test',
+    'pingpong reply = Whoopie Agoric!',
+  ]);
+});
+
+test.serial('plugin after restart', async t => {
+  const {
+    bridge,
+    cycle,
+    dump,
+    plugin,
+    queueThunkForKernel,
+  } = await setupVatController(t);
+
+  plugin.reset();
+  queueThunkForKernel(() => bridge.deliverInbound('pingpong'));
+  await cycle();
+
+  t.deepEqual(dump().log, [
+    'starting plugin test',
+    'installing pingPongP',
+    'starting pingpong test',
+    'pingpong reply = Whoopie Agoric!',
+    'starting pingpong test',
+    'pingpong reply = Whoopie Agoric!',
```

Comment:
and we should specifically *not* see a second copy here

Note that the `testLog` is ephemeral: it is not recorded in the kernelDB. However when vats are replayed, they *do* write their messages to the `testLog`, even though their syscalls are being simulated/blocked. So it kinda *looks* like the testLog is persistent, but it's not, it's just being deterministically regenerated.